### PR TITLE
Adds uniform processing kwargs to paligemma.

### LINKED ISSUE #31911: Uniform kwargs for processors
### Feature request

We want to standardize the logic flow through Processor classes. Since processors can have different kwargs depending on the model and modality, we are adding a `TypedDict` for each modality to keep track of which kwargs are accepted. 

The initial design is merged and an example model is modified to follow the new uniform processor kwargs in https://github.com/huggingface/transformers/pull/31198. Also https://github.com/huggingface/transformers/pull/31197 has two more examples with standardized API.

This design has to be shipped to all the processors in Transformers, and appreciate contributions. 
Below is an incomplete list of models that need standardization, feel free to add a model if it's missing:

- [ ] BLIP-2
- [ ] CLIP
- [ ] ClipSeg
- [ ] DepthAnything
- [ ] Dino
- [ ] Donut
- [ ] Fuyu
- [ ] Idefics
- [ ] Idefics-2
- [ ] InstructBlip
- [ ] InstructBlipVideo
- [ ] Kosmos-2
- [ ] LayoutLM (1, 2, 3)
- [ ] LLaVa
- [ ] LLaVa-NeXT
- [ ] LLaVa-NeXT-Video
- [ ] Maskformer
- [ ] Nougat
- [ ] Owlv2
- [ ] SigLip
- [ ] Paligemma
- [ ] VideoLLaVa
- [ ] VIT

Note: For now we'll start with image or image+text, https://github.com/huggingface/transformers/pull/31368 is an ongoing PR that has also audio processor standardization

### Motivation

.

### Your contribution

.

### PR DESCRIPTION
**Pull request by MnCSSJ4x**
# What does this PR do?
Adds Uniform Processing kwargs for paligemma model. 

Partially Fixes Issue 31911 

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Can Review - @zucchini-nlp

### CODE REVIEW

File: src/transformers/models/paligemma/processing_paligemma.py [hunk 1]
```diff
@@ -72,6 +74,18 @@ def build_string_from_input(prompt, bos_token, image_seq_len, image_token):
     """
     return f"{image_token * image_seq_len}{bos_token}{prompt}\n"
 
+class PaliGemmaProcessorKwargs(ProcessingKwargs, total=False):
+    _defaults = {
+        "text_kwargs": {
+            "tokenize_newline_separately": True,
+            "suffix": None,
```
Comment (zucchini-nlp) [inline src/transformers/models/paligemma/processing_paligemma.py:81]: We have to add all text-related kwargs to defaults if they were used in processor's call method. We're missing `padding`, `truncation` etc. Also, `suffix` is not present in general TextKwargs, so we have to add a PaligemmaTextKwargs. See more here
https://github.com/huggingface/transformers/blob/2af199c42b545f6248475ce456dd6c2a351b8522/src/transformers/processing_utils.py#L277-L291

File: src/transformers/models/paligemma/processing_paligemma.py [hunk 2]
```diff
@@ -72,6 +74,18 @@ def build_string_from_input(prompt, bos_token, image_seq_len, image_token):
...
+        },
+        "image_kwargs": {
+            "do_convert_rgb": None,
+            "do_thumbnail": None,
+            "do_align_long_axis":  None,
+        },
```
Comment (zucchini-nlp) [inline src/transformers/models/paligemma/processing_paligemma.py:87]: Same here, not all kwargs from call method are in defaults and might need PaligemmaImageKwargs if not all of them are in default ImageKwargs

File: src/transformers/models/paligemma/processing_paligemma.py [hunk 3]
```diff
@@ -253,27 +250,21 @@ def __call__(
...
-            image_mean=image_mean,
-            image_std=image_std,
-            input_data_format=input_data_format,
-            data_format=data_format,
-            resample=resample,
-            do_convert_rgb=do_convert_rgb,
+            **kwargs["image_kwargs"],
         )["pixel_values"]
 
+        max_length = kwargs.get("max_length", None)
```
Comment (zucchini-nlp) [inline src/transformers/models/paligemma/processing_paligemma.py:256]: not sure this will work, since it's not guaranteed `max_length` will be in kwargs and not `text_kwargs`. IMO we should first merge the kwargs and then try to get attributes like `suffix` or `max_length`

**Review by zucchini-nlp — COMMENTED** [review]
Hey @MnCSSJ4x ! Thanks for working on this! Left a few comments setting defaults values

Currently the tests are failing, so we have to fix them and make sure the paligemma_processing_tests contain the `ProcessorTesterMixin`. That is the way to ensure we're actually testing the new kwargs format. 

Also, I want to ask some paligemma-specific tests for processors as it seems that Paligemma has several model-specific kwargs 

cc @molbap if you have time to take a look

**Comment by zucchini-nlp** [comment]
Oh, can you remove the "Fixes #" from body. If not merging this PR will close the linked issue :)

File: src/transformers/models/paligemma/processing_paligemma.py [hunk 4]
```diff
@@ -74,19 +74,45 @@ def build_string_from_input(prompt, bos_token, image_seq_len, image_token):
...
+            "do_normalize": None,
+            "image_mean": None,
+            "image_std": None,
+            "data_format": "channels_first",
+            "input_data_format": None,
+            "resample": None,
+            "do_convert_rgb": None,  # Not Available in Default
+            "do_thumbnail": None,  # Not Available in Default
+            "do_align_long_axis": None,  # Not Available in Default
+            "do_rescale": None,
```
Comment (molbap) [inline src/transformers/models/paligemma/processing_paligemma.py:111]: nb: `None` defaults don't need to be included in `<MyModel>ProcessorKwargs` as it's the normal default!

File: src/transformers/models/paligemma/processing_paligemma.py [hunk 5]
```diff
@@ -74,19 +74,45 @@ def build_string_from_input(prompt, bos_token, image_seq_len, image_token):
...
+
+class PaliGemmaTextKwargs(TextKwargs):
+    suffix: Optional[Union[TextInput, PreTokenizedInput, List[TextInput], List[PreTokenizedInput]]] = None
+
+
+class PaliGemmaImagesKwargs(ImagesKwargs):
+    do_convert_rgb: Optional[bool] = None
+    do_thumbnail: Optional[bool] = None
+    do_align_long_axis: Optional[bool] = None
+    do_rescale: Optional[bool] = None
```
Comment (molbap) [inline src/transformers/models/paligemma/processing_paligemma.py:86]: `TypedDict` kwarg classes don't need to hold default `None` values or others, they only serve to indicate the accepted args and their type!

**Review by molbap — COMMENTED** [review]
Thanks for working on this! left a couple comments

**Comment by MnCSSJ4x** [comment]
@zucchini-nlp I changed the code to incorporate the new parameters. Let me know if I am going in the right direction. I haven't started writing the tests as I was occupied these few days. Will add that soon. Let me know if there is any reference which can help me understand the test or to take ideas from.

**Review by zucchini-nlp — COMMENTED** [review]
Great, thanks for re-iterating on this! For the tests you can take a look at this one: https://github.com/huggingface/transformers/blob/main/tests/models/align/test_processor_align.py. If paligemma doesn't have any processing test, you have to create a new file. Otherwise adding `ProcessorTesterMixin` will enable new tests for Paligemma. The general info about tests is [here](https://huggingface.co/docs/transformers/en/testing) :)

Additionally, can you add paligemma specific tests, which will test model-specific kwargs like `suffix` etc. When the CI turns greenm feel free to tag me and @molbap for review

File: src/transformers/models/paligemma/processing_paligemma.py [hunk 6]
```diff
@@ -19,17 +19,19 @@
 import logging
 from typing import List, Optional, Union
 
+
+try:
+    from typing import Unpack
+except ImportError:
+    pass
```
Comment (yonigozlan) [inline src/transformers/models/paligemma/processing_paligemma.py:26]:
```suggestion
if sys.version_info >= (3, 11):
    from typing import Unpack
else:
    from typing_extensions import Unpack
```
This will make the CI happier :)

File: src/transformers/models/paligemma/processing_paligemma.py [hunk 7]
```diff
@@ -253,27 +268,20 @@ def __call__(
...
-            input_data_format=input_data_format,
-            data_format=data_format,
-            resample=resample,
-            do_convert_rgb=do_convert_rgb,
+            **kwargs["image_kwargs"],
         )["pixel_values"]
-
+        max_length = output_kwargs.get("max_length", None)
         if max_length is not None:
             max_length += self.image_seq_length  # max_length has to account for the image tokens
```
Comment (yonigozlan) [inline src/transformers/models/paligemma/processing_paligemma.py:275]:
```suggestion
```

File: src/transformers/models/paligemma/processing_paligemma.py [hunk 8]
```diff
@@ -253,27 +268,20 @@ def __call__(
...
-
+        output_kwargs = self._merge_kwargs(
+            PaliGemmaProcessorKwargs,
+            tokenizer_init_kwargs=self.tokenizer.init_kwargs,
+            **kwargs,
+        )
```
Comment (yonigozlan) [inline src/transformers/models/paligemma/processing_paligemma.py:280]:
```suggestion
        if output_kwargs["text_kwargs"].get("max_length", None) is not None:
            output_kwargs["text_kwargs"]["max_length"] += self.image_seq_length
```
Instead of merging the kwargs again

File: src/transformers/models/paligemma/processing_paligemma.py (hunk 6 above)
Comment (MnCSSJ4x) [reply src/transformers/models/paligemma/processing_paligemma.py]: Thanks. Didn't know about sys.version_info.

**Comment by MnCSSJ4x** [comment]
Hey @yonigozlan I wrote some tests however got failure in tokenization and others where I checked the circle CI log and noticed gated repo issue. Is there a way to fix it or mock it up?

File: src/transformers/models/paligemma/processing_paligemma.py [hunk 9]
```diff
@@ -124,25 +152,8 @@ def __call__(
         self,
         text: Union[TextInput, PreTokenizedInput, List[TextInput], List[PreTokenizedInput]] = None,
         images: ImageInput = None,
```
Comment (yonigozlan) [inline src/transformers/models/paligemma/processing_paligemma.py:154]: These two inputs should be reversed and support for backward compatibility should be added. This should be similar to what is needed for Fuyu:
https://github.com/huggingface/transformers/blob/aa3bc0b4d5c98a40fc211994697d1893976f8bf9/src/transformers/models/fuyu/processing_fuyu.py#L522-L532

File: src/transformers/models/paligemma/processing_paligemma.py [hunk 10]
```diff
@@ -124,25 +152,8 @@ def __call__(
...
-        input_data_format: Optional[
-            Union[str, "ChannelDimension"]  # noqa: F821
-        ] = None,
-        resample: "PILImageResampling" = None,  # noqa: F821
-        do_convert_rgb: bool = None,
-        do_thumbnail: bool = None,
-        do_align_long_axis: bool = None,
-        do_rescale: bool = None,
-        suffix: Optional[Union[TextInput, PreTokenizedInput, List[TextInput], List[PreTokenizedInput]]] = None,
+        video=None,
```
Comment (yonigozlan) [inline src/transformers/models/paligemma/processing_paligemma.py:155]: `audio=None` Is still needed here for API consistency, even if this model doesn't support the audio modality.
````suggestion
        audio = None,
        video=None,

File: src/transformers/models/paligemma/processing_paligemma.py [hunk 11]
```diff
@@ -216,7 +227,12 @@ def __call__(
             - **pixel_values** -- Pixel values to be fed to a model. Returned when `images` is not `None`.
             - **labels** -- Labels compatible with training if `suffix` is not None
         """
-
+        output_kwargs = self._merge_kwargs(
+            PaliGemmaProcessorKwargs,
+            tokenizer_init_kwargs=self.tokenizer.init_kwargs,
+            **kwargs,
+        )
+        suffix = output_kwargs["text_kwargs"]["suffix"]
```
Comment (yonigozlan) [inline src/transformers/models/paligemma/processing_paligemma.py:235]: If suffix is not specified as a kwargs, this will cause an error. Better to use:
`suffix = output_kwargs["text_kwargs"].pop("suffix", None)`

File: src/transformers/models/paligemma/processing_paligemma.py [hunk 12]
```diff
@@ -73,6 +76,31 @@ def build_string_from_input(prompt, bos_token, image_seq_len, image_token):
...
+    do_align_long_axis: Optional[bool]
+    do_rescale: Optional[bool]
+
+
+class PaliGemmaProcessorKwargs(ProcessingKwargs, total=False):
+    text_kwargs: PaliGemmaTextKwargs
+    image_kwargs: PaliGemmaImagesKwargs
+    _defaults = {
+        "text_kwargs": {
+            "tokenize_newline_separately": True,
```
Comment (yonigozlan) [inline src/transformers/models/paligemma/processing_paligemma.py:95]: Looks like `tokenize_newline_separately` is not use anywhere, and it is not a default `text_kwargs`, so it might be best to remove it entirely?

File: src/transformers/models/paligemma/processing_paligemma.py [hunk 13]
```diff
@@ -253,27 +269,14 @@ def __call__(
...
+            output_kwargs["text_kwargs"]["max_length"] += self.image_seq_length
         inputs = self.tokenizer(
             input_strings,
             text_pair=suffix,
-            return_tensors=return_tensors,
-            padding=padding,
-            max_length=max_length,
-            truncation=truncation,
+            **output_kwargs["text_kwargs"],
             return_token_type_ids=return_token_type_ids,
```
Comment (yonigozlan) [inline src/transformers/models/paligemma/processing_paligemma.py:280]:
```suggestion
            return_token_type_ids=return_token_type_ids,
            **output_kwargs["text_kwargs"],
```

File: tests/models/paligemma/test_processing_paligemma.py [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+    )
+
+
+@require_vision
+class PaliGemmaProcessorTest(ProcessorTesterMixin, unittest.TestCase):
+    processor_class = PaliGemmaProcessor
+
+    def setUp(self):
+        self.tmpdirname = tempfile.mkdtemp()
+        processor = PaliGemmaProcessor.from_pretrained("google/paligemma-3b-pt-224")
```
Comment (yonigozlan) [inline tests/models/paligemma/test_processing_paligemma.py:42]: This will indeed cause a gated repo issue. you could rebuild a processor without using this repo, something like 
```suggestion
        image_processor = SiglipImageProcessor.from_pretrained("google/siglip-so400m-patch14-384")
        tokenizer = GemmaTokenizer(SAMPLE_VOCAB, keep_accents=True)
        processor = PaliGemmaProcessor(image_processor=image_processor, tokenizer=tokenizer)
```
Where
`SAMPLE_VOCAB = get_tests_dir("fixtures/test_sentencepiece.model")`
as it is done for `test_tokenization_gemma.py`.

Not sure if that's the nicest way to fix this though, any idea @zucchini-nlp @molbap ?

File: tests/models/paligemma/test_processing_paligemma.py [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+    # and size_divisor behaviour
+
+    @require_torch
+    @require_vision
+    def test_image_processor_defaults_preserved_by_image_kwargs(self):
+        if "image_processor" not in self.processor_class.attributes:
+            self.skipTest(f"image_processor attribute not present in {self.processor_class}")
+        image_processor = self.get_component(
+            "image_processor",
+            crop_size={"shortest_edge": 234, "longest_edge": 234},
```
Comment (yonigozlan) [inline tests/models/paligemma/test_processing_paligemma.py:75]: This may not be needed anymore as the base tests were changed recently, same for other tests. Please fetch and rebase on upstream main :)

**Review by yonigozlan — COMMENTED** [review]
Thanks for working on the tests! Left a few comments on the processor refactor also.
For the tests, as I indicated below, the base tests for `ProcessorTesterMixin` were recently changed. You can do this to have the newest changes on your branch:
`git fetch upstream`
`git rebase upstream/main`
You'll see that you may not have to override some of the tests at all :).

**Review by molbap — COMMENTED** [review]
Nice work! Let's see with the upstream rebase, I think we can reduce loc count by a fair chunk :hugs:

File: tests/models/paligemma/test_processing_paligemma.py (hunk 14 above)
Comment (ydshieh) [reply tests/models/paligemma/test_processing_paligemma.py:42]: Sorry, what is the issue here? This repo seem to be public no?

Comment (molbap) [reply tests/models/paligemma/test_processing_paligemma.py:42]: image
I suppose it is, with a license to accept?

Comment (ydshieh) [reply tests/models/paligemma/test_processing_paligemma.py:42]: wait, are we talking about 

google/paligemma-3b-pt-224

or 

google/siglip-so400m-patch14-384

but both are accessible even if I am using a firefox private window

File: tests/models/paligemma/test_processing_paligemma.py [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+    def prepare_image_inputs(self):
+        """This function prepares a list of PIL images, or a list of numpy arrays if one specifies numpify=True,
+        or a list of PyTorch tensors if one specifies torchify=True.
+        """
+
+        image_inputs = [np.random.randint(255, size=(3, 30, 400), dtype=np.uint8)]
+
+        image_inputs = [Image.fromarray(np.moveaxis(x, 0, -1)) for x in image_inputs]
+
+        return image_inputs
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply tests/models/paligemma/test_processing_paligemma.py:63]: Maybe this one works, I usually use it for image processor tests? `from tests.test_image_processing_common import prepare_image_inputs`

[the remark this answers is not in the document]
Comment (amyeroberts) [reply tests/models/paligemma/test_processing_paligemma.py:63]: Yep! Let's move. I also have my own personal agenda to remove the "numpify" and "torchify" arguments which are confusing, clash and inconsistent so would be a good opportunity for that
